MaxText/tests/attention_test.py

```diff
@@ -49,149 +46,172 @@ def setUp(self):
     self.mesh = Mesh(devices_array, self.cfg.mesh_axes)
 
     self.global_batch_size = self.cfg.global_batch_size_to_train_on
-    self.num_heads = self.cfg.base_num_heads
+    self.num_kv_heads = self.cfg.num_kv_heads
+    self.num_query_heads = self.cfg.num_query_heads
     self.max_target_length = self.cfg.max_target_length
     self.head_dim = self.cfg.head_dim
     self.embed_dim = self.cfg.base_emb_dim
     self.dtype = self.cfg.dtype
 
-    self.mha_attention = MultiHeadDotProductAttention(
-        num_heads=self.num_heads,
+    self._attention_as_mha_generic = Attention(
+        num_query_heads=self.num_query_heads,
+        num_kv_heads=self.num_kv_heads,
         head_dim=self.head_dim,
         mesh=self.mesh,
+        attention_kernel = "dot_product",
         dtype=self.dtype,
         dropout_rate=self.cfg.dropout_rate,
         name='self_attention',
     )
-    self.variable = self.mha_attention.init(
+
+    self._attention_as_mha_generic_variable = self._attention_as_mha_generic.init(
```

Comment:
Instead of using  ".init", , you may create params of q, k, v kernels with jnp.random and use flax.freeze for later.
i.e. variable = {} variable['params'] = {'query': ....,  'key', ... . 'value'...}
attention_layer.apply(freeze(variable), ....)

This would simplify your MQA test.

Reply:
Wow, I thought that `__init__` was shared but it isn't! Good to know!

Reply:
You could create mqa variable for such test only. Both ways shall work, `replace` is less readable to me, as `_attention_as_mha_generic_variable` is generated, it's structure is not very obvious.